go/vt/vttablet/tabletserver/semisyncwatcher/watcher.go

```diff
@@ -0,0 +1,282 @@
+		// we don't need to wait for anything.
+		return
+	}
+	// The primary is blocked. We need to wait for it to be unblocked.
+	ch := w.addWaiter()
+	<-ch
+}
+
+// stillBlocked returns true if the watcher should continue writing to the DB
+// because the watcher is still open, and the primary is still blocked.
+func (w *Watcher) stillBlocked() bool {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.isOpen && w.isBlocked
+}
+
+// checkAndSetIsWriting checks if the watcher is already writing to the DB.
+// If it is not, then it sets the isWriting field and signals the caller.
+func (w *Watcher) checkAndSetIsWriting() bool {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	if w.isWriting {
+		return false
+	}
+	w.isWriting = true
+	return true
+}
+
+// clearIsWriting clears the isWriting field.
+func (w *Watcher) clearIsWriting() {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.isWriting = false
+}
+
+// startWrites starts writing to the DB.
+// It is re-entrant and will return if we are already writing.
+func (w *Watcher) startWrites() {
+	// If we are already writing, then we can just return.
+	if !w.checkAndSetIsWriting() {
+		return
+	}
+	// We defer the clear of the isWriting field.
+	defer w.clearIsWriting()
+
+	// We start writing to the DB with a backoff.
+	backoff := 1 * time.Second
+	maxBackoff := 1 * time.Minute
+	// Check if we need to continue writing or not.
+	for w.stillBlocked() {
+		go w.write()
+		<-time.After(backoff)
+		backoff = time.Duration(math.Min(float64(backoff*2), float64(maxBackoff)))
+	}
+}
+
+// write writes a heartbeat to unblock semi-sync being stuck.
+func (w *Watcher) write() {
+	// Get a connection from the pool
+	conn, err := w.appPool.Get(context.Background())
+	if err != nil {
+		return
+	}
+	defer conn.Recycle()
+	_, _ = conn.Conn.ExecuteFetch(semiSyncRecoverWrite, 0, false)
+}
+
+// setIsBlocked sets the isBlocked field.
+func (w *Watcher) setIsBlocked(val bool) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.isBlocked = val
+	if !val {
+		// If we are unblocked, then we need to signal all the waiters.
+		for _, ch := range w.waiters {
+			close(ch)
+		}
+		// We also empty the list of current waiters.
+		w.waiters = nil
+	}
+}
+
+// clearAllData clears all the data in the table so that it never
+// consumes too much space on the MySQL instance.
+func (w *Watcher) clearAllData() {
+	// Get a connection from the pool
+	conn, err := w.appPool.Get(context.Background())
+	if err != nil {
+		log.Errorf("SemiSync Watcher: failed to clear semisync_recovery table: %v", err)
+		return
+	}
+	defer conn.Recycle()
+	_, err = conn.Conn.ExecuteFetch(semiSyncRecoverClear, 0, false)
+	if err != nil {
+		log.Errorf("SemiSync Watcher: failed to clear semisync_recovery table: %v", err)
+	}
+}
+
+// addWaiter adds a waiter to the list of waiters
+// that will be unblocked when the primary is no longer blocked.
+func (w *Watcher) addWaiter() chan any {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	ch := make(chan any, 1)
```

Comment:
We don't want this to be buffered, do we?

Reply:
Good catch. No we don't need a buffered channel.